src/lib/Templates.ml

```diff
@@ -120,74 +146,298 @@ let rec instantiate_instructions (param_map : var_map) : parsed_seq_code_bloc ->
+      let e2' = expand_parsed_discrete_boolean_expression synt_arrays e2 in
+      Parsed_comparison (e1', op, e2')
+  | Parsed_comparison_in (e1, e2, e3) ->
+      let e1' = expand_parsed_discrete_arithmetic_expression synt_arrays e1 in
+      let e2' = expand_parsed_discrete_arithmetic_expression synt_arrays e2 in
+      let e3' = expand_parsed_discrete_arithmetic_expression synt_arrays e3 in
+      Parsed_comparison_in (e1', e2', e3')
+  | Parsed_nested_bool_expr e -> Parsed_nested_bool_expr (expand_parsed_boolean_expression synt_arrays e)
+  | Parsed_not e -> Parsed_not (expand_parsed_boolean_expression synt_arrays e)
+
+and expand_parsed_discrete_arithmetic_expression synt_arrays = function
+  | Parsed_sum_diff (e, t, sum_diff) ->
+      let e' = expand_parsed_discrete_arithmetic_expression synt_arrays e in
+      let t' = expand_parsed_discrete_term synt_arrays t in
+      Parsed_sum_diff (e', t', sum_diff)
+  | Parsed_term t -> Parsed_term (expand_parsed_discrete_term synt_arrays t)
+
+and expand_parsed_discrete_term synt_arrays = function
+  | Parsed_product_quotient (t, f, product_quotient) ->
+      let t' = expand_parsed_discrete_term synt_arrays t in
+      let f' = expand_parsed_discrete_factor synt_arrays f in
+      Parsed_product_quotient (t', f', product_quotient)
+  | Parsed_factor f -> Parsed_factor (expand_parsed_discrete_factor synt_arrays f)
+
+and expand_parsed_discrete_factor synt_arrays = fun factor ->
+  let get_name_of_factor = function
+    | Parsed_variable (name, _) -> name
+    | _ -> failwith "[expand_parsed_discrete_factor]: Name of syntatic array was not a variable name."
   in
-  let rename_action_transition (guard, bloc, sync, loc_name) =
-    let sync' = rename_action_sync sync in
-    (guard, bloc, sync', loc_name)
+  match factor with
+    | Parsed_access (factor', index) -> begin
+        let arr_name = get_name_of_factor factor' in
+        match List.find_map (fun (name, _, len) -> if arr_name = name then Some len else None) synt_arrays with
+          | None -> factor
+          | Some len ->
+              let index_c = NumConst.to_bounded_int (get_const_disc_arith_expr index) in
+              if index_c < len then
+                let var_name = gen_access_id arr_name index_c in
+                Parsed_variable (var_name, 0)
+              else failwith "[expand_parsed_discrete_factor]: Index is greater or equal to length of syntatic array."
+    end
+    | _ -> factor
+
+let expand_name_or_access_list = List.map expand_name_or_access
+
+let expand_sync : unexpanded_sync -> sync = function
+  | UnexpandedSync action -> Sync (expand_name_or_access action)
+  | UnexpandedNoSync -> NoSync
+
+let expand_indexed_update (synt_vars : synt_vars_data) : parsed_scalar_or_index_update_type -> parsed_scalar_or_index_update_type =
+  fun e ->
+    match e with
+      | Parsed_indexed_update (Parsed_scalar_update (name, _), index) -> begin
```

Comment:
This code is almost the same as lines 276-285. Maybe it could be refactored ?

Reply:
Yes. I added a function `gen_var_from_access` and did the refactor with it. Let me know what do you think.

Reply:
nice :D